### PR TITLE
Introduce Form Groups

### PR DESCRIPTION
**Pull request by djhi**
Introducing form groups will allow us to remove the restrictions on the `FormTab` content which could accept any components and not just inputs while still tracking whether any of their inputs are invalid.

- [x] Base implementation
- [x] Cleanup
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: examples/simple/src/posts/PostEdit.js [hunk 1]
```diff
@@ -43,12 +45,25 @@ const EditActions = ({ basePath, data, hasShow }) => (
...
 
+const SanitizedBox = ({ fullWidth, basePath, ...props }) => <Box {...props} />;
+
 const PostEdit = ({ permissions, ...props }) => (
     <Edit title={<PostTitle />} actions={<EditActions />} {...props}>
         <TabbedForm initialValues={{ average_note: 0 }} warnWhenUnsavedChanges>
             <FormTab label="post.form.summary">
-                <TextInput disabled source="id" />
-                <TextInput source="title" validate={required()} resettable />
+                <SanitizedBox
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostEdit.js:54]: why do you need that? It seems unrelated to the PR.

File: packages/ra-core/src/form/FormContextProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+import { ReactNode } from 'react';
+import { FormContextValue } from '../types';
+import { FormContext } from './FormContext';
+
+/**
+ * Provides utilities to Forms children, allowing them to change the default save function or register inputs inside a group.
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormContextProvider.tsx:7]:
```suggestion
 * Provides utilities to Form children, allowing them to change the default save function or register inputs to a group.
```

File: packages/ra-core/src/form/FormGroupContextProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+export const FormGroupContextProvider = ({
+    children,
+    name,
+}: {
+    children: ReactNode;
+    name: string;
+}) => {
+    const formContext = useFormContext();
+
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContextProvider.tsx:22]: You should check that the formContext contains the `register` and `unregister` functions and throw a meaningful error if it doesn't. 

I'm surprised that TypeScript didn't warn you on this one.

File: packages/ra-core/src/form/FormGroupContext.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { createContext } from 'react';
+
+export const FormGroupContext = createContext<FormGroupContextValue>(undefined);
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContext.ts:3]: you should add a jsdoc saying that this shouldn't be used in standalone, but through FormGroupContextProvider

File: packages/ra-core/src/form/FormGroupContextProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import { ReactNode, useEffect } from 'react';
+import { FormGroupContext } from './FormGroupContext';
+import { useFormContext } from './useFormContext';
+
+/**
+ * This provider allows its input children to know they belong to a specific group.
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContextProvider.tsx:7]:
```suggestion
 * This provider allows its input children to register to a specific group.
```

File: packages/ra-core/src/form/FormGroupContextProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * This enables other components to access groups properties such as its
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContextProvider.tsx:8]:
```suggestion
 * This enables other components in the group to access group properties such as its
```

File: packages/ra-core/src/form/FormGroupContextProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * validation (valid/invalid) or whether its inputs have been updated (dirty/pristine).
+ * @param props The component props
+ * @param {ReactNode} props.children The form group content
+ * @param {String} props.name The form group name
+ */
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContextProvider.tsx:13]: Please add the same example you added to the doc here

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 8]
```diff
@@ -92,7 +93,38 @@ const FormWithRedirect: FC<FormWithRedirectProps> = ({
...
+            unregisterGroup: name => {
+                delete formGroups[name];
+            },
+            registerField: (source, group) => {
+                if (group) {
+                    formGroups.current[group] = formGroups.current[group] || [];
+                    const fields = new Set(formGroups.current[group]);
+                    fields.add(source);
+                    formGroups.current[group] = Array.from(fields);
+                }
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx:112]: what if group isn't defined? Shouldn't you throw an error?

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 9]
```diff
@@ -92,7 +93,38 @@ const FormWithRedirect: FC<FormWithRedirectProps> = ({
...
+            registerGroup: name => {
+                formGroups.current[name] = formGroups.current[name] || [];
+            },
+            unregisterGroup: name => {
+                delete formGroups[name];
+            },
+            registerField: (source, group) => {
+                if (group) {
+                    formGroups.current[group] = formGroups.current[group] || [];
+                    const fields = new Set(formGroups.current[group]);
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx:109]: you could merge the previous line with this one using

```
const fields = new Set(formGroups.current[group] | []);
```

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 10]
```diff
@@ -92,7 +93,38 @@ const FormWithRedirect: FC<FormWithRedirectProps> = ({
...
+            unregisterField: (source, group) => {
+                if (group) {
+                    if (!formGroups.current[group]) {
+                        console.warn(`Invalid form group ${group}`);
+                    } else {
+                        const fields = new Set(formGroups.current[group]);
+                        fields.delete(source);
+                        formGroups.current[group] = Array.from(fields);
+                    }
+                }
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx:123]: same, what happens when called without a group?

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 11]
```diff
@@ -92,7 +93,38 @@ const FormWithRedirect: FC<FormWithRedirectProps> = ({
...
+            getGroupFields: name => formGroups.current[name] || [],
+            registerGroup: name => {
+                formGroups.current[name] = formGroups.current[name] || [];
+            },
+            unregisterGroup: name => {
+                delete formGroups[name];
+            },
+            registerField: (source, group) => {
+                if (group) {
+                    formGroups.current[group] = formGroups.current[group] || [];
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx:108]: why do you allow an input to register to a group that hasn't been registered here, while you forbid it in the following function?

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 12]
```diff
@@ -92,7 +93,38 @@ const FormWithRedirect: FC<FormWithRedirectProps> = ({
         [save]
     );
 
-    const formContextValue = useMemo(() => ({ setOnSave }), [setOnSave]);
+    const formContextValue = useMemo<FormContextValue>(
+        () => ({
+            setOnSave,
+            getGroupFields: name => formGroups.current[name] || [],
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx:99]: As these functions are completely static, I suggest you define them outside of the component.

File: packages/ra-core/src/form/useFormContext.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { useContext } from 'react';
+import { FormContext } from './FormContext';
+
+/**
+ * Retrieve the form context enabling consumers to alter its save function or to register inputs inside a form group.
+ * @returns {FormContext} The form context.
+ */
+export const useFormContext = () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormContext.ts:8]: could be a one liner

File: packages/ra-core/src/form/useFormGroup.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    pristine: boolean;
+    dirty: boolean;
+};
+
+/**
+ * Retrieve a specific form group data such as its validation status (valid/invalid) or
+ * or whether its inputs have been updated (dirty/pristine)
+ * @param {string] name The form group name
+ * @returns {FormGroupState} The form group state
+ */
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormGroup.ts:19]: please add a usage example

File: packages/ra-core/src/form/useFormGroup.ts [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    const formContext = useFormContext();
+    const [state, setState] = useState<FormGroupState>({
+        errors: undefined,
+        valid: true,
+        invalid: false,
+        pristine: true,
+        dirty: false,
+    });
+
+    form.subscribe(
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormGroup.ts:31]: you never unsubscribe... I think there's a leak here

File: packages/ra-core/src/form/useFormGroup.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        () => {
+            const fields = formContext.getGroupFields(name);
+            const newState = Array.from(fields).reduce<FormGroupState>(
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormGroup.ts:34]: why the Array.from()?

File: packages/ra-core/src/form/useFormGroup.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+                (acc, field) => {
+                    const fieldState = form.getFieldState(field);
+                    let errors = acc.errors;
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormGroup.ts:37]: `let errors = acc.errors || [];` and remove line 40 to 42

File: packages/ra-core/src/form/useInput.ts [hunk 18]
```diff
@@ -44,6 +46,19 @@ const useInput = ({
     ...options
 }: InputProps): UseInputValue => {
     const finalName = name || source;
+    const formGroupName = useFormGroupContext();
+    const formContext = useFormContext();
+
+    useEffect(() => {
+        if (!formContext) {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useInput.ts:53]:
```suggestion
        if (!formContext || !formGroupName ) {
```

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx [hunk 19]
```diff
@@ -44,7 +53,9 @@ describe('<SaveButton />', () => {
             <TestContext>
                 <ThemeProvider theme={theme}>
                     <SaveContextProvider value={saveContextValue}>
-                        <SaveButton {...invalidButtonDomProps} />
+                        <FormContextProvider value={formContextValue}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:56]: I don't understand why this is now required and it worries me (that it's now required). this will be a gotcha for users.

File: packages/ra-core/src/form/useFormGroup.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+                    if (fieldState.error) {
+                        if (!errors) {
+                            errors = {};
+                        }
+                        errors[field] = fieldState.error;
+                    }
+
+                    const newState = {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useFormGroup.ts:46]: Please add a unit test for that logic

File: packages/ra-ui-materialui/src/form/TabbedForm.tsx [hunk 21]
```diff
@@ -344,7 +339,14 @@ const sanitizeRestProps = ({
     ...props
 }) => props;
 
+export default TabbedForm;
+
 export const findTabsWithErrors = (children, errors) => {
+    // TODO: Provide documentation link
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.tsx:345]: todo

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Great work! The error detection code in the tab component is much more readable. 

I have the general feeling that this could be more unit tested.

File: examples/simple/src/posts/PostEdit.js (hunk 1 above)
Comment (djhi) [reply examples/simple/src/posts/PostEdit.js:54]: It was a test to verify that we could now have anything inside a `FormTab` and the validation would still work. I can revert it if you want

File: packages/ra-core/src/form/FormWithRedirect.tsx (hunk 8 above)
Comment (djhi) [reply packages/ra-core/src/form/FormWithRedirect.tsx:112]: I lazily initialize it here at L108. I found out that inputs were registering themselves before the group anyway.

File: packages/ra-core/src/form/FormWithRedirect.tsx (hunk 10 above)
Comment (djhi) [reply packages/ra-core/src/form/FormWithRedirect.tsx:123]: I don't get this comment: I do check whether a group was provided (L113) and it exists (L114)

File: packages/ra-core/src/form/FormWithRedirect.tsx (hunk 11 above)
Comment (djhi) [reply packages/ra-core/src/form/FormWithRedirect.tsx:108]: I found out that inputs were registering themselves before the group

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx (hunk 19 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:56]: I actually don't get how those tests could have pass before

File: packages/ra-core/src/form/FormGroupContextProvider.tsx [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    children,
+    name,
+}: {
+    children: ReactNode;
+    name: string;
+}) => {
+    const formContext = useFormContext();
+
+    useEffect(() => {
+        if (!formContext) {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormGroupContextProvider.tsx:60]:
```suggestion
        if (!formContext || !formContext.registerGroup || !formContext.unregisterGroup) {
```

File: packages/ra-core/src/form/FormWithRedirect.tsx (hunk 8 above)
Comment (fzaninotto) [reply packages/ra-core/src/form/FormWithRedirect.tsx:111]: I wasn't clear. My question was: what if `registerField` is called with an empty second argument? But I understand that it's the general case (no group declared) and that in this case there is nothing to do.

File: packages/ra-core/src/form/FormWithRedirect.tsx (hunk 10 above)
Comment (fzaninotto) [reply packages/ra-core/src/form/FormWithRedirect.tsx:122]: Nevermind.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Almost there!

**Review by fzaninotto — APPROVED** [review]
